File: shared/error_handling.py
"""
Centralized error handling utilities for MCP E-commerce Demo
Provides consistent error types, logging, and response formatting
"""

import logging
import traceback
from datetime import datetime
from enum import Enum
from typing import Dict, Any, Optional, Union
from dataclasses import dataclass


class ErrorCode(Enum):
    """Standard error codes across the application"""
    # General errors
    UNKNOWN_ERROR = "UNKNOWN_ERROR"
    VALIDATION_ERROR = "VALIDATION_ERROR"
    AUTHENTICATION_ERROR = "AUTHENTICATION_ERROR"
    AUTHORIZATION_ERROR = "AUTHORIZATION_ERROR"
    NOT_FOUND = "NOT_FOUND"
    RATE_LIMIT_EXCEEDED = "RATE_LIMIT_EXCEEDED"
    
    # MCP protocol errors (following JSON-RPC 2.0 spec)
    MCP_PARSE_ERROR = "MCP_PARSE_ERROR"  # -32700
    MCP_INVALID_REQUEST = "MCP_INVALID_REQUEST"  # -32600
    MCP_METHOD_NOT_FOUND = "MCP_METHOD_NOT_FOUND"  # -32601
    MCP_INVALID_PARAMS = "MCP_INVALID_PARAMS"  # -32602
    MCP_INTERNAL_ERROR = "MCP_INTERNAL_ERROR"  # -32603
    
    # Business logic errors
    CART_ERROR = "CART_ERROR"
    PRODUCT_ERROR = "PRODUCT_ERROR"
    CHECKOUT_ERROR = "CHECKOUT_ERROR"
    SESSION_ERROR = "SESSION_ERROR"
    
    # External service errors
    OPENAI_ERROR = "OPENAI_ERROR"
    NEKUDA_ERROR = "NEKUDA_ERROR"
    MCP_SERVER_ERROR = "MCP_SERVER_ERROR"


# Map error codes to HTTP status codes
ERROR_CODE_TO_HTTP_STATUS = {
    ErrorCode.UNKNOWN_ERROR: 500,
    ErrorCode.VALIDATION_ERROR: 400,
    ErrorCode.AUTHENTICATION_ERROR: 401,
    ErrorCode.AUTHORIZATION_ERROR: 403,
    ErrorCode.NOT_FOUND: 404,
    ErrorCode.RATE_LIMIT_EXCEEDED: 429,
    ErrorCode.MCP_PARSE_ERROR: 400,
    ErrorCode.MCP_INVALID_REQUEST: 400,
    ErrorCode.MCP_METHOD_NOT_FOUND: 404,
    ErrorCode.MCP_INVALID_PARAMS: 400,
    ErrorCode.MCP_INTERNAL_ERROR: 500,
    ErrorCode.CART_ERROR: 400,
    ErrorCode.PRODUCT_ERROR: 400,
    ErrorCode.CHECKOUT_ERROR: 400,
    ErrorCode.SESSION_ERROR: 400,
    ErrorCode.OPENAI_ERROR: 502,
    ErrorCode.NEKUDA_ERROR: 502,
    ErrorCode.MCP_SERVER_ERROR: 502,
}

# Map error codes to MCP JSON-RPC error codes
ERROR_CODE_TO_MCP_CODE = {
    ErrorCode.MCP_PARSE_ERROR: -32700,
    ErrorCode.MCP_INVALID_REQUEST: -32600,
    ErrorCode.MCP_METHOD_NOT_FOUND: -32601,
    ErrorCode.MCP_INVALID_PARAMS: -32602,
    ErrorCode.MCP_INTERNAL_ERROR: -32603,
    ErrorCode.UNKNOWN_ERROR: -32603,
    ErrorCode.VALIDATION_ERROR: -32602,
    ErrorCode.NOT_FOUND: -32601,
    ErrorCode.CART_ERROR: -32603,
    ErrorCode.PRODUCT_ERROR: -32603,
    ErrorCode.CHECKOUT_ERROR: -32603,
    ErrorCode.SESSION_ERROR: -32603,
    ErrorCode.OPENAI_ERROR: -32603,
    ErrorCode.NEKUDA_ERROR: -32603,
    ErrorCode.MCP_SERVER_ERROR: -32603,
}


@dataclass
class ErrorContext:
    """Additional context for error reporting"""
    user_id: Optional[str] = None
    session_id: Optional[str] = None
    request_id: Optional[str] = None
    operation: Optional[str] = None
    component: Optional[str] = None
    additional_data: Optional[Dict[str, Any]] = None


class AppError(Exception):
    """Base application error class with structured error information"""
    
    def __init__(
        self,
        message: str,
        error_code: ErrorCode = ErrorCode.UNKNOWN_ERROR,
        context: Optional[ErrorContext] = None,
        cause: Optional[Exception] = None,
        user_message: Optional[str] = None
    ):
        super().__init__(message)
        self.message = message
        self.error_code = error_code
        self.context = context or ErrorContext()
        self.cause = cause
        self.user_message = user_message or self._get_default_user_message()
        self.timestamp = datetime.utcnow()
        
    def _get_default_user_message(self) -> str:
        """Get user-friendly error message based on error code"""
        user_messages = {
            ErrorCode.VALIDATION_ERROR: "Invalid input provided. Please check your data and try again.",
            ErrorCode.NOT_FOUND: "The requested resource was not found.",
            ErrorCode.RATE_LIMIT_EXCEEDED: "Too many requests. Please wait and try again.",
            ErrorCode.CART_ERROR: "There was an issue with your cart. Please try again.",
            ErrorCode.PRODUCT_ERROR: "Product information is temporarily unavailable.",
            ErrorCode.CHECKOUT_ERROR: "Checkout failed. Please review your information and try again.",
            ErrorCode.SESSION_ERROR: "Your session has expired. Please refresh the page.",
            ErrorCode.OPENAI_ERROR: "AI service is temporarily unavailable. Please try again.",
            ErrorCode.NEKUDA_ERROR: "Payment service is temporarily unavailable. Please try again.",
            ErrorCode.MCP_SERVER_ERROR: "Service is temporarily unavailable. Please try again.",
        }
        return user_messages.get(self.error_code, "An unexpected error occurred. Please try again.")
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert error to dictionary for JSON serialization"""
        return {
            "error_code": self.error_code.value,
            "message": self.message,
            "user_message": self.user_message,
            "timestamp": self.timestamp.isoformat(),
            "context": {
                "user_id": self.context.user_id,
                "session_id": self.context.session_id,
                "request_id": self.context.request_id,
                "operation": self.context.operation,
                "component": self.context.component,
                "additional_data": self.context.additional_data,
            } if self.context else None,
            "cause": str(self.cause) if self.cause else None,
        }
    
    def get_http_status(self) -> int:
        """Get appropriate HTTP status code for this error"""
        return ERROR_CODE_TO_HTTP_STATUS.get(self.error_code, 500)
    
    def get_mcp_error_code(self) -> int:
        """Get appropriate MCP JSON-RPC error code"""
        return ERROR_CODE_TO_MCP_CODE.get(self.error_code, -32603)


class ValidationError(AppError):
    """Error for validation failures"""
    def __init__(self, message: str, context: Optional[ErrorContext] = None, cause: Optional[Exception] = None):
        super().__init__(message, ErrorCode.VALIDATION_ERROR, context, cause)


class NotFoundError(AppError):
    """Error for resource not found"""
    def __init__(self, message: str, context: Optional[ErrorContext] = None, cause: Optional[Exception] = None):
        super().__init__(message, ErrorCode.NOT_FOUND, context, cause)


class CartError(AppError):
    """Error for cart operations"""
    def __init__(self, message: str, context: Optional[ErrorContext] = None, cause: Optional[Exception] = None):
        super().__init__(message, ErrorCode.CART_ERROR, context, cause)


class CheckoutError(AppError):
    """Error for checkout operations"""
    def __init__(self, message: str, context: Optional[ErrorContext] = None, cause: Optional[Exception] = None):
        super().__init__(message, ErrorCode.CHECKOUT_ERROR, context, cause)


class ExternalServiceError(AppError):
    """Error for external service failures"""
    def __init__(
        self, 
        service_name: str, 
        message: str, 
        context: Optional[ErrorContext] = None, 
        cause: Optional[Exception] = None
    ):
        error_code = {
            "openai": ErrorCode.OPENAI_ERROR,
            "nekuda": ErrorCode.NEKUDA_ERROR,
            "mcp_server": ErrorCode.MCP_SERVER_ERROR,
        }.get(service_name.lower(), ErrorCode.UNKNOWN_ERROR)
        
        super().__init__(f"{service_name} error: {message}", error_code, context, cause)


def setup_structured_logging(component: str, level: str = None) -> logging.Logger:
    """Setup structured logging for a component with environment-based configuration"""
    import os
    
    # Get log level from environment or use provided level or default to INFO
    if level is None:
        level = os.getenv("LOG_LEVEL", "INFO").upper()
    
    # Get environment type
    environment = os.getenv("ENVIRONMENT", "development").lower()
    
    logger = logging.getLogger(component)
    
    if not logger.handlers:
        handler = logging.StreamHandler()
        
        # Different formatting for production vs development
        if environment == "production":
            # Cleaner format for production
            formatter = logging.Formatter(
                '%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                datefmt='%Y-%m-%d %H:%M:%S'
            )
        else:
            # More detailed format for development
            formatter = logging.Formatter(
                '%(asctime)s | %(levelname)8s | %(name)15s | %(filename)s:%(lineno)d | %(message)s',
                datefmt='%Y-%m-%d %H:%M:%S'
            )
        
        handler.setFormatter(formatter)
        logger.addHandler(handler)
    
    logger.setLevel(getattr(logging, level.upper(), logging.INFO))
    
    # Log configuration on first setup
    if component not in getattr(setup_structured_logging, '_configured_components', set()):
        if not hasattr(setup_structured_logging, '_configured_components'):
            setup_structured_logging._configured_components = set()
        setup_structured_logging._configured_components.add(component)
        logger.info(f"Logging configured for {component}: level={level}, environment={environment}")
    
    return logger


def log_error(
    logger: logging.Logger,
    error: Union[AppError, Exception],
    context: Optional[ErrorContext] = None
) -> None:
    """Log an error with structured information"""
    if isinstance(error, AppError):
        error_data = error.to_dict()
        logger.error(
            f"[{error.error_code.value}] {error.message}",
            extra={
                "error_code": error.error_code.value,
                "context": error_data.get("context"),
                "cause": error_data.get("cause"),
                "user_message": error.user_message,
            }
        )
        if error.cause:
            logger.error(f"Caused by: {error.cause}")
            if hasattr(error.cause, '__traceback__'):
                logger.debug("Traceback: %s", traceback.format_exception(
                    type(error.cause), error.cause, error.cause.__traceback__
                ))
    else:
        context_info = ""
        if context:
            context_parts = []
            if context.operation:
                context_parts.append(f"op={context.operation}")
            if context.session_id:
                context_parts.append(f"session={context.session_id}")
            if context.component:
                context_parts.append(f"component={context.component}")
            if context_parts:
                context_info = f" [{', '.join(context_parts)}]"
        
        logger.error(f"Unexpected error{context_info}: {error}")
        logger.debug("Traceback: %s", traceback.format_exception(
            type(error), error, error.__traceback__
        ))


def handle_error_recovery(
    error: Exception,
    operation: str,
    fallback_action: Optional[callable] = None,
    logger: Optional[logging.Logger] = None
) -> Any:
    """Handle error recovery with optional fallback action"""
    if logger:
        context = ErrorContext(operation=operation)
        log_error(logger, error, context)
    
    if fallback_action:
        try:
            return fallback_action()
        except Exception as fallback_error:
            if logger:
                logger.error(f"Fallback action failed for {operation}: {fallback_error}")
            raise error  # Re-raise original error
    
    raise error


def create_error_response(error: Union[AppError, Exception], request_id: Optional[str] = None) -> Dict[str, Any]:
    """Create standardized error response"""
    if isinstance(error, AppError):
        return {
            "error": {
                "code": error.get_mcp_error_code(),
                "message": error.user_message,
                "data": {
                    "error_code": error.error_code.value,
                    "details": error.message,
                    "timestamp": error.timestamp.isoformat(),
                    "request_id": request_id,
                }
            }
        }
    else:
        return {
            "error": {
                "code": -32603,
                "message": "An unexpected error occurred",
                "data": {
                    "error_code": "UNKNOWN_ERROR",
                    "details": str(error),
                    "timestamp": datetime.utcnow().isoformat(),
                    "request_id": request_id,
                }
            }
        }

File: shared/error_handling.ts
/**
 * Centralized error handling utilities for MCP E-commerce Demo Frontend
 * Provides consistent error types, logging, and response formatting
 */

export enum ErrorCode {
  // General errors
  UNKNOWN_ERROR = 'UNKNOWN_ERROR',
  VALIDATION_ERROR = 'VALIDATION_ERROR',
  AUTHENTICATION_ERROR = 'AUTHENTICATION_ERROR',
  AUTHORIZATION_ERROR = 'AUTHORIZATION_ERROR',
  NOT_FOUND = 'NOT_FOUND',
  RATE_LIMIT_EXCEEDED = 'RATE_LIMIT_EXCEEDED',
  NETWORK_ERROR = 'NETWORK_ERROR',
  
  // Business logic errors
  CART_ERROR = 'CART_ERROR',
  PRODUCT_ERROR = 'PRODUCT_ERROR',
  CHECKOUT_ERROR = 'CHECKOUT_ERROR',
  SESSION_ERROR = 'SESSION_ERROR',
  
  // External service errors
  API_ERROR = 'API_ERROR',
  MCP_ERROR = 'MCP_ERROR',
  NEKUDA_ERROR = 'NEKUDA_ERROR',
  
  // UI errors
  RENDER_ERROR = 'RENDER_ERROR',
  COMPONENT_ERROR = 'COMPONENT_ERROR',
}

export interface ErrorContext {
  userId?: string
  sessionId?: string
  requestId?: string
  operation?: string
  component?: string
  url?: string
  additionalData?: Record<string, any>
}

export class AppError extends Error {
  public readonly errorCode: ErrorCode
  public readonly context: ErrorContext
  public readonly cause?: Error
  public readonly userMessage: string
  public readonly timestamp: Date

  constructor(
    message: string,
    errorCode: ErrorCode = ErrorCode.UNKNOWN_ERROR,
    context: ErrorContext = {},
    cause?: Error,
    userMessage?: string
  ) {
    super(message)
    this.name = 'AppError'
    this.errorCode = errorCode
    this.context = context
    this.cause = cause
    this.userMessage = userMessage || this.getDefaultUserMessage()
    this.timestamp = new Date()
  }

  private getDefaultUserMessage(): string {
    const userMessages: Record<ErrorCode, string> = {
      [ErrorCode.VALIDATION_ERROR]: 'Invalid input provided. Please check your data and try again.',
      [ErrorCode.NOT_FOUND]: 'The requested resource was not found.',
      [ErrorCode.RATE_LIMIT_EXCEEDED]: 'Too many requests. Please wait and try again.',
      [ErrorCode.NETWORK_ERROR]: 'Network connection issue. Please check your connection and try again.',
      [ErrorCode.CART_ERROR]: 'There was an issue with your cart. Please try again.',
      [ErrorCode.PRODUCT_ERROR]: 'Product information is temporarily unavailable.',
      [ErrorCode.CHECKOUT_ERROR]: 'Checkout failed. Please review your information and try again.',
      [ErrorCode.SESSION_ERROR]: 'Your session has expired. Please refresh the page.',
      [ErrorCode.API_ERROR]: 'Service is temporarily unavailable. Please try again.',
      [ErrorCode.MCP_ERROR]: 'Communication error. Please try again.',
      [ErrorCode.NEKUDA_ERROR]: 'Payment service is temporarily unavailable. Please try again.',
      [ErrorCode.RENDER_ERROR]: 'Display error occurred. Please refresh the page.',
      [ErrorCode.COMPONENT_ERROR]: 'Component error occurred. Please try again.',
      [ErrorCode.UNKNOWN_ERROR]: 'An unexpected error occurred. Please try again.',
      [ErrorCode.AUTHENTICATION_ERROR]: 'Authentication failed. Please log in again.',
      [ErrorCode.AUTHORIZATION_ERROR]: 'You do not have permission to perform this action.',
    }
    return userMessages[this.errorCode] || userMessages[ErrorCode.UNKNOWN_ERROR]
  }

  toJSON(): Record<string, any> {
    return {
      errorCode: this.errorCode,
      message: this.message,
      userMessage: this.userMessage,
      timestamp: this.timestamp.toISOString(),
      context: this.context,
      cause: this.cause?.message,
    }
  }
}

export class ValidationError extends AppError {
  constructor(message: string, context: ErrorContext = {}, cause?: Error) {
    super(message, ErrorCode.VALIDATION_ERROR, context, cause)
  }
}

export class NetworkError extends AppError {
  constructor(message: string, context: ErrorContext = {}, cause?: Error) {
    super(message, ErrorCode.NETWORK_ERROR, context, cause)
  }
}

export class CartError extends AppError {
  constructor(message: string, context: ErrorContext = {}, cause?: Error) {
    super(message, ErrorCode.CART_ERROR, context, cause)
  }
}

export class CheckoutError extends AppError {
  constructor(message: string, context: ErrorContext = {}, cause?: Error) {
    super(message, ErrorCode.CHECKOUT_ERROR, context, cause)
  }
}

export class APIError extends AppError {
  public readonly status?: number

  constructor(message: string, status?: number, context: ErrorContext = {}, cause?: Error) {
    super(message, ErrorCode.API_ERROR, context, cause)
    this.status = status
  }
}

// Logging levels
export enum LogLevel {
  DEBUG = 'debug',
  INFO = 'info',
  WARN = 'warn',
  ERROR = 'error',
}

// Structured logger interface
export interface Logger {
  debug(message: string, context?: Record<string, any>): void
  info(message: string, context?: Record<string, any>): void
  warn(message: string, context?: Record<string, any>): void
  error(message: string, context?: Record<string, any>): void
}

class ConsoleLogger implements Logger {
  constructor(private component: string, private level: LogLevel = LogLevel.INFO) {}

  private shouldLog(level: LogLevel): boolean {
    const levels = [LogLevel.DEBUG, LogLevel.INFO, LogLevel.WARN, LogLevel.ERROR]
    return levels.indexOf(level) >= levels.indexOf(this.level)
  }

  private formatMessage(level: LogLevel, message: string, context?: Record<string, any>): void {
    if (!this.shouldLog(level)) return

    const timestamp = new Date().toISOString()
    const prefix = `${timestamp} | ${level.toUpperCase().padEnd(5)} | ${this.component.padEnd(15)} |`
    
    if (context && Object.keys(context).length > 0) {
      console[level](`${prefix} ${message}`, context)
    } else {
      console[level](`${prefix} ${message}`)
    }
  }

  debug(message: string, context?: Record<string, any>): void {
    this.formatMessage(LogLevel.DEBUG, message, context)
  }

  info(message: string, context?: Record<string, any>): void {
    this.formatMessage(LogLevel.INFO, message, context)
  }

  warn(message: string, context?: Record<string, any>): void {
    this.formatMessage(LogLevel.WARN, message, context)
  }

  error(message: string, context?: Record<string, any>): void {
    this.formatMessage(LogLevel.ERROR, message, context)
  }
}

// Logger factory
export function createLogger(component: string, level: LogLevel = LogLevel.INFO): Logger {
  return new ConsoleLogger(component, level)
}

// Error logging utility
export function logError(logger: Logger, error: Error | AppError, context?: ErrorContext): void {
  if (error instanceof AppError) {
    logger.error(`[${error.errorCode}] ${error.message}`, {
      context: error.context,
      cause: error.cause?.message,
      userMessage: error.userMessage,
      ...context,
    })
  } else {
    logger.error(`Unexpected error: ${error.message}`, {
      name: error.name,
      stack: error.stack,
      ...context,
    })
  }
}

// Error recovery utility
export async function handleErrorRecovery<T>(
  operation: () => Promise<T>,
  operationName: string,
  fallbackAction?: () => Promise<T>,
  logger?: Logger
): Promise<T> {
  try {
    return await operation()
  } catch (error) {
    if (logger) {
      const context: ErrorContext = { operation: operationName }
      logError(logger, error as Error, context)
    }

    if (fallbackAction) {
      try {
        logger?.info(`Attempting fallback for ${operationName}`)
        return await fallbackAction()
      } catch (fallbackError) {
        logger?.error(`Fallback action failed for ${operationName}`, {
          fallbackError: (fallbackError as Error).message,
        })
        throw error // Re-throw original error
      }
    }

    throw error
  }
}

// Retry utility with exponential backoff
export async function retryWithBackoff<T>(
  operation: () => Promise<T>,
  options: {
    maxRetries?: number
    baseDelay?: number
    maxDelay?: number
    operationName?: string
    logger?: Logger
  } = {}
): Promise<T> {
  const {
    maxRetries = 3,
    baseDelay = 1000,
    maxDelay = 10000,
    operationName = 'operation',
    logger,
  } = options

  let lastError: Error

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      if (attempt > 0) {
        logger?.info(`Retry attempt ${attempt} for ${operationName}`)
      }
      return await operation()
    } catch (error) {
      lastError = error as Error
      
      if (attempt === maxRetries) {
        logger?.error(`All retry attempts failed for ${operationName}`, {
          attempts: attempt + 1,
          finalError: lastError.message,
        })
        throw lastError
      }

      const delay = Math.min(baseDelay * Math.pow(2, attempt), maxDelay)
      logger?.warn(`${operationName} failed, retrying in ${delay}ms`, {
        attempt: attempt + 1,
        error: lastError.message,
      })
      
      await new Promise(resolve => setTimeout(resolve, delay))
    }
  }

  throw lastError!
}

// Error boundary helper for Vue components
export function createErrorHandler(component: string, logger?: Logger) {
  return (error: Error, info: string) => {
    const appError = new AppError(
      error.message,
      ErrorCode.COMPONENT_ERROR,
      { component, additionalData: { info } },
      error
    )
    
    if (logger) {
      logError(logger, appError)
    } else {
      console.error(`[${component}] Component Error:`, error, info)
    }
  }
}

// Notification helper for user-facing errors
export interface NotificationOptions {
  type: 'error' | 'warning' | 'info' | 'success'
  title?: string
  message: string
  duration?: number
  action?: {
    label: string
    handler: () => void
  }
}

export function createErrorNotification(error: Error | AppError): NotificationOptions {
  if (error instanceof AppError) {
    return {
      type: 'error',
      title: 'Error',
      message: error.userMessage,
      duration: 5000,
    }
  } else {
    return {
      type: 'error',
      title: 'Unexpected Error',
      message: 'An unexpected error occurred. Please try again.',
      duration: 5000,
    }
  }
}
